plugins/analytics-module-ga4/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts

```diff
@@ -0,0 +1,269 @@
+          return {
+            type: c.getString('type') as CustomDimensionOrMetricConfig['type'],
+            index: c.getNumber('index'),
+            source: c.getString(
+              'source',
+            ) as CustomDimensionOrMetricConfig['source'],
+            key: c.getString('key'),
+          };
+        }) ?? [];
+
+    if (identity === 'required' && !options.identityApi) {
+      throw new Error(
+        'Invalid config: identity API must be provided to deps when ga.identity is required',
+      );
+    }
+
+    // Return an implementation instance.
+    return new GoogleAnalytics({
+      ...options,
+      identity,
+      measurementId,
+      virtualSearchPageView,
+      cdmConfig,
+      testMode,
+    });
+  }
+
+  /**
+   * Primary event capture implementation. Handles core navigate event as a
+   * pageview and the rest as custom events. All custom dimensions/metrics are
+   * applied as they should be (set on pageview, merged object on events).
+   */
+  captureEvent(event: AnalyticsEvent) {
+    const { context, action, subject, value, attributes } = event;
+    const customMetadata = this.getCustomDimensionMetrics(context, attributes);
+
+    if (action === 'navigate' && context.extension === 'App') {
+      this.capture.pageview(subject, customMetadata);
+      return;
+    }
+
+    if (this.virtualSearchPageView.mode !== 'disabled' && action === 'search') {
+      const { mountPath, searchQuery, categoryQuery } =
+        this.virtualSearchPageView;
+      const params = new URLSearchParams();
+      params.set(searchQuery, subject);
+      if (categoryQuery) {
+        params.set(categoryQuery, context.searchTypes?.toString() ?? '');
+      }
+      this.capture.pageview(
+        `${mountPath}?${params.toString()}`,
+        customMetadata,
+      );
+      if (this.virtualSearchPageView.mode === 'only') {
+        return;
+      }
+    }
+
+    this.capture.event({
+      category: context.extension || 'App',
+      action,
+      label: subject,
+      value,
+      ...customMetadata,
+    });
+  }
+
+  /**
+   * Returns an object of dimensions/metrics given an Analytics Context and an
+   * Event Attributes, e.g. { dimension1: "some value", metric8: 42 }
+   */
+  private getCustomDimensionMetrics(
+    context: AnalyticsContextValue,
+    attributes: AnalyticsEventAttributes = {},
+  ) {
+    const customDimensionsMetrics: { [x: string]: string | number | boolean } =
+      {};
+
+    this.cdmConfig.forEach(config => {
+      const value =
+        config.source === 'context'
+          ? context[config.key]
+          : attributes[config.key];
+
+      // Never pass a non-numeric value on a metric.
+      if (config.type === 'metric' && typeof value !== 'number') {
+        return;
+      }
+
+      // Only set defined values.
+      if (value !== undefined) {
+        customDimensionsMetrics[`${config.type}${config.index}`] = value;
+      }
+    });
+
+    return customDimensionsMetrics;
+  }
```

Comment:
The purpose of this code is to map backstage analytics event attribute and context keys to UA (Universal Analytics) custom dimensions and metrics.  This was previously necessary because custom dimensions were not _named_ in UA, they just corresponded to numbered slots (e.g. `dimension1` or `metric3`)

In GA4 however, custom dimensions and metrics _are_ named, so I think there's opportunity to simplify this (and associated configurations) dramatically.

Rather than requiring admins to configure and map attributes/context keys, we could either:

- Not have a config at all, and instead rely on the fact that admins would have to configure them in the GA UI anyway, for the data to appear in reports.  We could update the GA code to dump all custom attributes/context keys onto the sent payload.
- Or have a much simpler `allowedContext` and `allowedAttributes` list configuration, which controlled which values were added to event payloads.